### PR TITLE
Closes #656: Add find in page functionality to engine/session

### LINKED ISSUE #656: Engine/Session: Find in page
This issue tracks the work needed to add "find in page" functionality to our `browser-engine-*` components and exposing the state of find in page (keyword, no of items found, etc.) in the `browser-session` component. This issue does **not** cover any UI work that may be implemented by a `feature-*` component.

This feature is primarily used by Focus (#200).

### PR DESCRIPTION
**Pull request by csadilek**

### CODE REVIEW

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -160,6 +160,32 @@ class GeckoEngineSession(
...
+            GeckoResult<Void>()
+        }
+        notifyObservers { onFind(text) }
+    }
+
+    /**
+     * See [EngineSession.findNext]
+     */
+    override fun findNext(forward: Boolean) {
+        val findFlags = if (forward) 0 else GeckoSession.FINDER_FIND_BACKWARDS
```
[the remark this answers is not in the document]
Comment (jonalmeida) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Nevermind, I looked it up. There isn't.

**Comment by csadilek** [comment]
> Aren't we missing a clear method?

Yes, we do :). I'll update the PR.

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (csadilek) [reply components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Yes, looked twice for this as well. Thanks for the review!

**Comment by csadilek** [comment]
@pocmo updated now. Can you take another look?

File: components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 2]
```diff
@@ -157,6 +157,27 @@ class GeckoEngineSession(
...
+     */
+    override fun findNext(forward: Boolean) {
+        // no-op (requires v63+)
+    }
+
+    /**
+     * See [EngineSession.clearFindMatches]
+     */
+    override fun clearFindMatches() {
+        // no-op (requires v63+)
```
Comment (pocmo) [inline components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Isn't that 62+? It looks like you implemented that in beta for 62 :)

File: components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 3]
```diff
@@ -164,6 +164,39 @@ class GeckoEngineSession(
...
+     * See [EngineSession.findAll]
+     */
+    override fun findAll(text: String) {
+        geckoSession.finder.find(text, 0).then { result: GeckoSession.FinderResult? ->
+            result?.let {
+                notifyObservers { onFindResult(it.current, it.total, true) }
+            }
+            GeckoResult<Void>()
+        }
+        notifyObservers { onFind(text) }
```
Comment (pocmo) [inline components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Can there be a race condition here? onFind() seems to clear the list of results in `EngineObserver`. Can it be possible that `onFindResult` gets called before `onFind` here? Should we call `onFind` first?

**Pull request merged**